File: services/catalog/ingest.py
"""One-off ingest: load movies into Mongo.

    python ingest.py                              # bundled sample (16 titles)
    python ingest.py --dataset data/tmdb_5000.csv # TMDB 5000

Idempotent: upserts keyed on movieId, safe to re-run.
"""
from __future__ import annotations

import argparse
import asyncio
import json
import sys
from pathlib import Path

from pymongo import UpdateOne

from app import db
from app.csv_util import parse_tmdb_csv


def _default_dataset() -> Path:
    # /data is the canonical in-container location (compose mounts ./data there).
    p = Path("/data/sample_movies.json")
    if p.exists():
        return p
    return Path(__file__).resolve().parent.parent.parent / "data" / "sample_movies.json"


async def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--dataset", help="path to TMDB CSV; omit for the bundled sample")
    args = ap.parse_args()

    if args.dataset and args.dataset.endswith(".csv"):
        records = [m for m in parse_tmdb_csv(Path(args.dataset).read_text()) if m["overview"]]
        print(f"parsed {len(records)} rows from {args.dataset}")
    else:
        path = Path(args.dataset) if args.dataset else _default_dataset()
        records = json.loads(path.read_text())
        print(f"loaded {len(records)} rows from {path}")

    if not records:
        print("no movies to ingest", file=sys.stderr)
        sys.exit(1)

    await db.connect()
    ops = [UpdateOne({"movieId": m["movieId"]}, {"$set": m}, upsert=True) for m in records]
    res = await db.movies().bulk_write(ops, ordered=False)
    print(f"ingest complete: upserted={res.upserted_count} modified={res.modified_count}")
    await db.close()


if __name__ == "__main__":
    asyncio.run(main())

File: services/catalog/src/ingest.ts
// One-off ingest job: loads movies into Mongo. Reads the bundled sample
// dataset by default, or a TMDB 5000 CSV when --dataset is passed.
//   npm run ingest                          # sample_movies.json (16 titles)
//   npm run ingest -- --dataset data/tmdb_5000.csv
import { readFileSync } from "node:fs";
import { connect, movies, close, Movie } from "./db.js";
import { parseTmdbCsv } from "./csv.js";
import { logger } from "./logger.js";

function parseArgs(argv: string[]): { dataset?: string } {
  const out: { dataset?: string } = {};
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--dataset") out.dataset = argv[++i];
  }
  return out;
}

async function main(): Promise<void> {
  const { dataset } = parseArgs(process.argv.slice(2));
  await connect();

  let records: Movie[];
  if (dataset && dataset.endsWith(".csv")) {
    records = parseTmdbCsv(readFileSync(dataset, "utf8")).filter((m) => m.overview);
    logger.info({ dataset, count: records.length }, "parsed TMDB CSV");
  } else {
    const path = dataset ?? new URL("../../../data/sample_movies.json", import.meta.url).pathname;
    records = JSON.parse(readFileSync(path, "utf8"));
    logger.info({ path, count: records.length }, "loaded sample dataset");
  }

  // Bulk upsert keyed on movieId — idempotent, safe to re-run.
  const ops = records.map((m) => ({
    updateOne: { filter: { movieId: m.movieId }, update: { $set: m }, upsert: true },
  }));
  if (ops.length) await movies().bulkWrite(ops, { ordered: false });
  logger.info({ inserted: ops.length }, "ingest complete");
  await close();
}

main().catch((err) => {
  logger.error({ err }, "ingest failed");
  process.exit(1);
});
